### ISSUE TITLE
DCM S3

### ISSUE THREAD

**Issue by pameyer** [IQSS/dataverse#4703]
The DCM (data capture module, for big data upload) currently integrate with Datavese assumes POSIX storage (and ssh+rsync data transfer); however using object stores (AWS S3, OpenStack Swift, etc) is becoming more common.

Open technical design questions:
- S3 (and swift?) are sometimes considered transfer protocols in addition to storage protocols.  Should a S3 DCM support these as data transfer protocols, or only as storage?
- DCM design assumes that having client side checksums calculated without direct user intervention is essential to "big data depositions", and that these checksums should be propagated from deposition to publication (and data file replication, etc).  How do other disciplines view this trade-off (implementation complexity vs data integrity)?

Lowest-way complexity way of implementing this would be a second DCM implementation (or configuration option for current DCM), changing only how data files are transferred from the temporary upload location to the dataverse-accessable storage (aka - internal copy from temporary POSIX to DV S3/swift dataset buckets) and keeping existing ssh+rsync and client-side checksums.  Moderately higher complexity would involve changes to the existing approach for client-side checksums and data transfers to support non-unix OS systems.

**Comment by pameyer** [comment]
Notes from discussion:
- support S3/swift as transfer protocols too? no
- support swift storage too? no, separate issue
- go with low-complexity approach (aka - change scanner to use S3 commands for `/upload` -> `/hold`
- @landreev pointed out that there could be issues with the batch import (from the DCM success API); current mapping of `/hold` to `$GLASSFISH_DIR/glassfish/domains/domain1/files` may need review / possible revisiting.

**Comment by matthew-a-dunlap** [comment]
My understanding of the work needed for the story:
[ ] Copy/update data-capture-module's scanner to use S3 instead of posix
[ ] Update Datasets.java receiveChecksumValidationResults to not refer to the local file storage. Namely, call a new command instead of ImportFromFileSystemCommand()

**Comment by pameyer** [comment]
One open question: should updating the existing DCM jenkins tests to include S3 be added to the list?

**Comment by matthew-a-dunlap** [comment]
After discussing with @pameyer , my understanding is that this work does not include functionality to allow downloading of files uploaded via DCM S3. The end result of this work will be that the file-bundle is uploaded to a structured S3 bucket that dataverse is aware of. This will lay the foundation for the later download functionality.

**Comment by matthew-a-dunlap** [comment]
Here is some important info about Dataverse's batch file system jobs https://github.com/sbgrid/sbgrid-dataverse/wiki/INFO:-Batch-Job-Implementation https://github.com/sbgrid/sbgrid-dataverse/wiki/HOWTO:-Batch-Job-Testing#3-manual-tests .

Posting these here for future developers as well as myself. The functionality around this needs to at least be understood when implementing an s3 version of the import.

Talking with @landreev , we were unsure as to whether this s3 functionality should live as a new batch job or if it could somehow be rolled into our existing storageIO code. If possible, the latter seems simpler from a code maintenance perspective.

**Comment by matthew-a-dunlap** [comment]
Another facet: should we consider creating an actual package (e.g. tar/zip) to store on s3? AWS has per-file charges for getting/moving files around in buckets, and if we aren't providing the ability to download individual files in the dcm-dataset anyways, maybe we should consider actually zipping things up?

**Comment by matthew-a-dunlap** [comment]
- [ ] Copy/update data-capture-module's scanner to use S3 instead of posix. This means calling aws commands within the bash script.
- [ ] Update Datasets.java receiveChecksumValidationResults to not refer to the local file storage. Namely, call a new command instead of ImportFromFileSystemCommand().
- [ ] Decide whether DCM jenkins tests should be updated, and if so update them.
   - This may also involve updating jenkins to run from a local docker

- [ ] More changes to #4761, which was blocking this story?

**Comment by matthew-a-dunlap** [comment]
I've identified an issue with setting up docker-dcm where the Dataverse installer does not complete when being run in non-interactive mode. It looks to be specifically related to glassfish-startup, where the installer does not get past this part:
`When/if asadmin asks you to "Enter admin user name",
it should be safe to hit return and accept the default
(which is "admin").

Press any key to continue...`
There may be some new installer option I'm missing. This is the line that is running the installer for dcm: `./install -admin_email=dvAdmin@mailinator.com -y -f > install.out 2> install.err`
This may have come about after the work in https://github.com/IQSS/dataverse/issues/4764 ?

**Comment by matthew-a-dunlap** [comment]
The above issue does not block the dcm s3 work but makes it a bit more annoying to get docker-dcm up and running (as you must run each step in /opt/dv/install.bash manually). 

My understand of this story is that I need to document the steps needed to have dcm use s3 instead of posix, but that actually automating that in docker is not needed at this point. I am pretty close to having that be deliverable but need to go over it a bit more tomorrow.

**Comment by pdurbin** [comment]
> It looks to be specifically related to glassfish-startup

This is a new script added by @thaorell and @tkmonson and I just mentioned your trouble to them at https://github.com/IQSS/dataverse/issues/4874#issuecomment-408432993

**Comment by matthew-a-dunlap** [comment]
With this story, testing is needed both on the posix rsync updates and the rsync-s3 code.There are a lot of manual steps needed to get the environment up and running for testing. I will post these up in a few parts.

**Comment by matthew-a-dunlap** [comment]
Initially setting up the docker images:
- Install docker if you do not have it
- Use the 4703-merge-dual branch
- Follow along with docker-dcm/readme.txt
   - This has you first do steps in docker-aio
   - Note that when you are waiting for the images to show up as healthy, glassfish takes like 5 minutes even if you see a healthy status. Definitely wait.

**Comment by matthew-a-dunlap** [comment]
To run rsync posix (also needed for rsync s3):
    * DataCite instead of ezid
        * connect to dvsrv box: docker exec -it dvsrv /bin/bash
        * run curl: curl -X PUT -d DataCite "http://localhost:8080/api/admin/settings/:DoiProvider"
    * Update dcm with branch code (rpm pulled by dataverse is out of date and incompatible)
        * Files
            * api/sr.py
                * docker cp /Users/matthew/Documents/github/data-capture-module/api/sr.py dcmsrv:/opt/dcm/api/sr.py
            * gen/usg.py
                * docker cp /Users/matthew/Documents/github/data-capture-module/gen/usg.py dcmsrv:/opt/dcm/gen/usg.py
            * scn/post_upload.bash
                * docker cp /Users/matthew/Documents/github/data-capture-module/scn/post_upload.bash dcmsrv:/opt/dcm/scn/post_upload.bash
            * 
            * ansible/roles/dcm/files/root/scripts/send_fail.bash
                * unused?
            * ansible/roles/dcm/files/root/scripts/send_ok.bash
                * unused?

**Comment by matthew-a-dunlap** [comment]
- To run rsync s3
   - Before: the default bucket for dcm to hold files in s3 is named test-dcm. It is coded into post_upload.bash (line 30). If you want a different bucket then use it.
      - This bucket should exist on our testing environment, but if it is not there go and create it through the browser ui
   - docker cp /Users/matthew/Documents/github/data-capture-module/scn/post_upload_s3.bash dcmsrv:/opt/dcm/scn/post_upload.bash
   - Install aws on dcmsrv and symlink it
      - pip install awscli --upgrade --user
      - export PATH=~/.local/bin:$PATH
   - Configure aws bucket on dcmsrv
      - you need a credentials files in ~/.aws
         - mkdir ~/.aws
         - yum install nano
         - nano ~/.aws/credentials
            - Structure like:
               - [default]
               - aws_access_key_id = 
               - aws_secret_access_key =
   - Dataverse configuration:
      - Set s3 as the storage driver
         - Log into the dcm docker box
         - cd /opt/glassfish4/bin/
         - ./asadmin delete-jvm-options "\-Ddataverse.files.storage-driver-id=file"
         - ./asadmin create-jvm-options "\-Ddataverse.files.storage-driver-id=s3"
      - Set up aws configs for dataverse
         - mkdir ~/.aws
         - yum install nano
         - nano ~/.aws/credentials
            - Structure like:
               - [default]
               - aws_access_key_id = 
               - aws_secret_access_key =
         - ALSO: create a region file
            - nano ~/.aws/config
            - contents:
               - [default]
               - region = us-east-1
      - Add the bucket names to dataverse
         - s3 bucket
         - s3 bucket for dcm (as dataverse needs to do the copy over)
            - ./asadmin create-jvm-options "-Ddataverse.files.s3-bucket-name=iqsstestdcmbucket"
            - ./asadmin create-jvm-options "-Ddataverse.files.dcm-s3-bucket-name=test-dcm"
            - iqsstestdcmbucket

**Comment by matthew-a-dunlap** [comment]
Actual operation (normal and s3):
- Create dataset and download rsync upload script
- Upload script to dcm_client (if needed, you can probably do all the actions for create/download inside dcm_client)
   - docker cp ~/Downloads/upload-FK2_NN49YM.bash dcm_client:/tmp
- Create a folder of files to upload (you just need something...)
- Run script
   - bash ./upload-FK2_NN49YM.bash
- manually run post_upload.bash on dcmsrv
   - bash ./opt/dcm/scn/post_upload.bash

**Comment by matthew-a-dunlap** [comment]
**Some additional tips on tweaking / developing in the docker containers:**
- You can completely blow away all the docker images with these commands (including non dcm ones!)
   - docker stop dvsrv
   - docker stop dcm_client
   - docker stop dcmsrv
   - docker rm $(docker ps -a -q)
   - docker rmi $(docker images -q)
- There are a few logs to tail
   - dvsrv : server.log
   - dcmsrv : tail -n 2000 -f /var/log/lighttpd/breakage.log
   - dcmsrv : tail -n 2000 -f /var/log/lighttpd/access.log
- I have attached a script for redeploying dataverse into your docker image if you just want to update the code. You'll probably need to tweak the paths [deploy_dataverse_into_docker.sh.zip](https://github.com/IQSS/dataverse/files/2237096/deploy_dataverse_into_docker.sh.zip)

**Comment by matthew-a-dunlap** [comment]
Because this code was built off the #4610 branch which is still in QA, I have not created a normal pull request yet for this issue. Instead refer to this compare for the differences https://github.com/IQSS/dataverse/compare/4610-upload-dual-mode...4703-merge-dual

Some of this work was done on the DCM side, please refer to this pr for that: https://github.com/sbgrid/data-capture-module/pull/28

Lastly, the steps for deploying are somewhat a part of this issue, even if we defer putting them in better docs for now.

**Comment by matthew-a-dunlap** [comment]
New To-Do:
- [ ] Pete's requested DCM side changes https://github.com/sbgrid/data-capture-module/pull/28
- [ ] Pull out dual mode from DCM S3

**Comment by pdurbin** [comment]
It was non-obvious for me how to find the instructions on the DCM side for configuring S3 stuff so I made a pull request: https://github.com/sbgrid/data-capture-module/pull/31 . I don't have a strong opinion of how it's worded but someone landing on the git repo for DCM should be able to click something in the readme or installation instructions to find out how to configure DCM for S3.

**Comment by pdurbin** [comment]
Pull request #4946 looks good to me. Moving to QA at https://waffle.io/IQSS/dataverse

**Comment by landreev** [comment]
@pdurbin has beat me to moving the PR to QA. 
But I re-reviewed it yesterday too, and I concur, that it looks ready.

**Comment by pameyer** [comment]
Ran through some initial testing with DCM 0.2 (POSIX) and docker-dcm; used dataverseAdmin to create datasets with UI and API.  Locking still seems to be behaving.  Engineered checksum failures result in 2x notifications to dataverseAdmin; checksum success messages result in 3x notifications to dataverseAdmin (some of this may be due to either configuration or user error).

**Comment by pameyer** [comment]
Ran into a non-intuitive failure mode during testing: It was possible to "half-create" a dataset, where a given dataset would show up when browsing (because it was present in solr), but clicking on the dataset gave a 404 (because it was not present in postgres).  Lighttpd logs showed requests from dataverse for `sr.py` but not `ur.py`.

Initial investigation suggested possible relation with dataverse role assignments.  More in-depth debugging showed that root cause was incorrectly configured DCM (specifically, adapting ansible role designed to deploy from `git pull` to the new `rpm` installation has incorrect CGI configuration, resulting in the DCM returning the source code of `ur.py` instead of the specified `{"status":"OK"}`.  Dataverse was checking HTTP status code (which was 200), but not the contents of the result.

Short version - Dataverse should probably be more rigorous in what it expects to receive from integration systems, but "happy path" is not impacted.

**Comment by pdurbin** [comment]
It seems to be easy to replicate the 3x notifications to the curator by running the InReviewWorkflowIT tests logging in as the curator:

![screen shot 2018-08-21 at 1 50 50 pm](https://user-images.githubusercontent.com/21006/44419503-61fd8380-a549-11e8-8718-0c48f1c055f9.png)

**Comment by pdurbin** [comment]
@pameyer and I just chatted about my "bad news" test in DatasetsIT where I construct some JSON with `badNews.add("status", "validation failed")`. I showed that notifications seemed to be working ok. The superuser got a single notification. The author got a single notification. Pete is going to dig in and see if he can help me replicate the bug.

**Comment by pameyer** [comment]
On `4703-dcm-s3-2-2f91725`, the only duplicate (checksum success or checksum failure) notifications I'm seeing are to administrative users (not the normal test users).  So it's either fixed, or was user/configuration error to begin with.

**Comment by pdurbin** [comment]
@pameyer ok. I take it there's nothing for me to fix, then. Thanks for testing! I'm taking myself off this issue.

**Comment by kcondon** [comment]
Found 2 issues:
1. aws cli should be installed as part of DCM setup
2. Why 2 buckets? config instructions indicate a dataverse and a dcm bucket but files only appear in dataverse bucket.

**Comment by pameyer** [comment]
As of 8d702eb6bbf0908efd0063243996d464950d7596; AWS CLI is installed into the DCM container for docker-dcm.

**Comment by pdurbin** [comment]
For the checksum fail notification I fixed the icon in 23ba120 and change the DOI to the dataset title with a link in c5e6298. @djbrooke looked over my shoulder.

**Comment by pdurbin** [comment]
I can tell from the code that the second bucket (``dataverse.files.dcm-s3-bucket-name``) is used and not cruft but I don't feel confident in my ability to document it properly. I added a FIXME in 5649d05 and a little text that should be improved upon by someone who wrote the code or someone who is testing the S3 support for DCM works.

**Comment by pdurbin** [comment]
@kcondon and I have been talking for over an hour about this issue and drew this:

![dcms3](https://user-images.githubusercontent.com/21006/44602622-1d265680-a7ae-11e8-955c-1eecb0aaf60a.png)

The additional change we agreed to I just made in d359f73 where I tried to explain a little more in the dev guide about what the new "dcm-s3" bucket is for. I also linked back to this issue because it turns out there's useful stuff in here but the GitHub web interface hid it from my because there are so many comments. I added a FIXME to document all of this stuff properly in the guides some day. It's extremely complex. Diagrams would be helpful as well.

Moving to QA, mostly to test the checksum notification improvements I made today and mentioned above.
